src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php

```diff
@@ -331,9 +331,26 @@ private function set($type, $id)
      *
      * @throws RuntimeException
      */
-    private function createAutowiredDefinition(\ReflectionClass $typeHint)
+    private function createAutowiredDefinition(\ReflectionClass $typeHint, Definition $decorator = null)
     {
         if (isset($this->ambiguousServiceTypes[$typeHint->name])) {
+            // Only two alternatives, check if it is a decorator/inner couple to use the decorator
```

Comment:
Yeah, more than 1 level of decoration is not handled right now. To be honest I don't see how to do it properly and I think handling 1 level would cover most cases